### ISSUE TITLE
flicker under high load

### ISSUE THREAD

**Issue by scandum** [mintty/mintty#1098]
When mintty is processing a lot of text it has a tendency to briefly display black bars between screen updates.

This is most obvious when repeatedly printing text outside the scrolling region.

It looks like mintty clears the line, next updates it, but under high load the update gets paused in the middle, resulting in flicker.

The obvious solution is to not do any screen clearing and to simply overwrite, would be sufficient if this was only done when printing outside the scroll region.

**Comment by mintty** [comment]
Overwriting only is the normal strategy of mintty. However, there are some exceptions with a separate background clearing action, for example when displaying italic text which may overlap in neighbour cells.
Also, if an application clears a line explicitly, this may become effective by one display cycle before the application rewrites the line, more likely under high load. There isn't much mintty could do in the latter case, unless perhaps, all clearing would be delayed by some "smart" strategy. Questionable effort, though.

It's hard to tell what effect you are really seeing. Please provide a test case.

**Comment by scandum** [comment]
It's actually proving tricky to provide a simple test case. I do have a complex test case but I want to give it a closer look.

Would it be possible to introduce an escape code that causes mintty to wait with printing text until the full buffer is available?

I guess it could be like \e7 and \e8, just a simple start and end code.

**Comment by mintty** [comment]
If it's your own application anyway (so you could add such sequence), why don't you simply drop the line clearing?
Test case:
[testcase.zip](https://github.com/mintty/mintty/files/6194409/testcase.zip)
Call it with a "-" parameter to run without clear-to-eol.

**Comment by avih** [comment]
> Would it be possible to introduce an escape code that causes mintty to wait with printing text until the full buffer is available?

Full buffer of what? The terminal can't know automatically that the application now considers the buffer "full".

However, there is a suggested escape sequence for that - synchronized update, where the application tells the terminal "suspend screen updates from now until I tell you to unsupend it" exactly to avoid tearing and flicker:
https://gitlab.com/gnachman/iterm2/-/wikis/synchronized-updates-spec

**Comment by scandum** [comment]
> If it's your own application anyway (so you could add such sequence), why don't you simply drop the line clearing?
> Test case:
> [testcase.zip](https://github.com/mintty/mintty/files/6194409/testcase.zip)
> Call it with a "-" parameter to run without clear-to-eol.

Upon closer examination it's indeed clearing related. The cursor getting stuck between updates does cause flicker, but I can fix that by disabling the cursor while printing.

As avih mentioned it looks like iTerm2 introduced a sync sequence. I'm not a fan of the particular sequence that was picked however. I'd rather see the introduction of a simpler alternative.

mintty performs flawless as far as I can tell, and I think I can work around the issue, so I'm closing this.

**Comment by mintty** [comment]
> Sync=\EP=%p1%ds\E\\

Where's that described for tmux? It's inconsistent with the spec proposal quoted before; the spec suggests a Begin and an End sequence while `Sync` specifies an embedding sequence. Either way, I think using DCS (ESC P) is calling for trouble, so if I'd ever take this up, I guess I'd rather go with an OSC or even CSI/DECSET sequence.

**Comment by avih** [comment]
> Where's that described for tmux?

I didn't check the docs, so I'm only referring to the source code (all links are to current tmux master - be568ea3).

As far as I understand, that's the sequence it uses by default if you only specify the capability by name (`Sync`) (and it's not at the current terminfo?):
https://github.com/tmux/tmux/blob/f5b7ebc540fe3db5c3a5fe97d01de44551e76058/tty-features.c#L183-L187

> It's inconsistent with the spec proposal quoted before; the spec suggests a Begin and an End sequence while Sync specifies an embedding sequence

I'm not very versed in terminfo terminology (no pun intended), but as far as I understand it is consistent - it invokes/prints the same capability with a different parameter - `1` for start, `2` for end:
Start: https://github.com/tmux/tmux/blob/be568ea3b21b88763d2cc274bdec3a476136de2f/tty.c#L1453
End: https://github.com/tmux/tmux/blob/be568ea3b21b88763d2cc274bdec3a476136de2f/tty.c#L1468

**Comment by BrianInglis** [comment]
This should be handled by the screen update optimizer. If slow scroll is not enabled, and all $LINES have been updated since the last refresh (or in the last refresh interval, or in the last $LINES*$COLUMNS/($BAUD/10) seconds < ~80ms [assuming 300x80@3Mb/s - I have stty 3000000 for years]), clear the client area and render the glyphs in the character locations. Nicht wahr?

**Comment by avih** [comment]
> This should be handled by the screen update optimizer...

Logic to automate (guesstimate) screen update timing without explicit synchronization sequences is definitely useful to have, however:
- It's not always simple to assess.
- There will ALWAYS be cases which it fails to guess correctly.
- In my experience, mintty is too slow to make such timing assessments reasonably useful (I know, I implemented such timing system in st, and experimented with such approach in mintty too, though I didn't necessarily exhaust the experimentations).

**Comment by mintty** [comment]
Hallo Brian,
slow scroll is not supported (I once tried to implement it but failed miserably; just getting another idea how it could be done, but does anybody really want it?).
I can follow you to the idea to postpone the next output cycle but by what criteria should mintty decide when to flush delayed output? Full screen having been updated is probably the rarest case.
@avih, how is mintty too slow? In my experience it's in fact one of the fastest terminals, maybe except st indeed. (Unless you're running cygwin 3.1.* which slows down pty transfer.)

**Comment by avih** [comment]
> how is mintty too slow? In my experience it's in fact one of the fastest terminals

I don't recall the details right now.

My vague recollection is that the rendering is slow, i.e. the function which updates the window to the new (internal) terminal state. IIRC it's the same function which does all the font selection etc.

I think when I measured it was in the order of several ms, and, depending on content, could even be frequently more than 16.7 ms. That was on a reasonably fast system, though not state of the art.

Considering that any automatic timing system will almost certainly need to operate with much less than 16.7ms timing margins, I concluded back than it's not fast enough for such system.

These experiments were performed probably about two year ago, but I don't recall the exact cygwin version.

IIRC I collected the timing data using `clock_gettime(CLOCK_MONOTONIC, &now);`.
I don't recall whether or not I used the win32 `timeBeginPeriod` to increase the timer resolution.

And of course, I might have deduced wrongly. I'd be glad to be proven wrong.

**Comment by scandum** [comment]
As for criteria, to get decent rendering with mintty under load you have to disable the cursor while printing.

So the re-enabling of the cursor `\e[?25h` could be used as a universal sync marker.

**Comment by avih** [comment]
> So the re-enabling of the cursor `\e[?25h` could be used as a universal sync marker.

There are many parameters which can improve the guessing, or make it worse. Also, what happens if the application doesn't re-show the cursor? So some timeout needs to be applied, and under load all the timings get compressed, and then it becomes even harder to know ehen to render, considering that the cursor is only visible some of the time, but the terminal cannot know in advance whether it should wait for it or just render now because $reasons.

It's not impossible, but it's also certainly not trivial to get reasonably useful behavior under wide spectrum of use cases.

**Comment by mintty** [comment]
I'm attaching an experimental patch. Suspend output sequence is `ESC[`N`&q` which sets up N ms delay, with no upper bound yet.
I don't know why the flushing function of the reset sequence (N=0) does not work without the additional usleep delay.
Test script included.
[suspend-test.zip](https://github.com/mintty/mintty/files/6203385/suspend-test.zip)

**Comment by avih** [comment]
> I'm attaching an experimental patch

```diff
+      if (arg0 == 0) {
+        do_update();
+        usleep(1000);  // flush update (??)
+      }
```

I recommend measuring how long it actually sleeps, especially when mintty is the only application running (close the browser and any other application).

The default windows timer resolution is 15.6 ms, so on average I imagine this sleep would be half of that, at which case may reduce performance meaningfully when this sequence is used.

Also, I don't think you should introduce your own sequence for that (except for testing) when an RFC with a sequence already exists. But that's up to you.

**Comment by scandum** [comment]
> Also, I don't think you should introduce your own sequence for that (except for testing) when an RFC with a sequence already exists. But that's up to you.

If you're talking about the iterm2 sequence, I personally wouldn't implement it because it's the wrong sequence for the job it's supposed to do.

> I don't know why the flushing function of the reset sequence (N=0) does not work without the additional usleep delay.

0 values in timers can be tricky. Looks like it works, though I was expecting each `\e[..&q` sequence to dual-function as a flush point. Still some oddities.

I'll attach the source code I use so you can take a look for yourself. gcc -O3 quadsort.c to compile. Need to zoom out the terminal to 100+ rows for it to display right while running. 

[quadsort.zip](https://github.com/mintty/mintty/files/6205344/quadsort.zip)

**Comment by avih** [comment]
> If you're talking about the iterm2 sequence

iTerm2, and st (with a patch) and tmux for about a year now, and patch for mpv exists (`--vo-tct`), and alacritty is considering it. And these are only the ones I'm aware of. I'm not following all terminals.

**Comment by scandum** [comment]
Random idea, slightly related.

How about an escape sequence that stops mintty from updating the screen, with a second escape sequence that causes mintty to go back to normal and do a screen update. This would be useful if you want to cat a few megabytes of text and are not interested in watching it scroll by, but also for certain types of animations.

Also a good way to benchmark the performance irrespective of display updates.

**Comment by BrianInglis** [comment]
That's a side effect of <ctrl-C> INTR and the function of <ctrl-O> DISCARD which should be restored to the Cygwin DLL console code soon; hopefully also working again will be<ctrl-\> QUIT.

**Comment by mintty** [comment]
> n escape sequence that stops mintty from updating the screen, with a second escape sequence that causes mintty to go back to normal and do a screen update

Isn't that just the sequence we've been discussing? I've implemented it meanwhile, with the addition that a timeout can be given as a second parameter: `ESC P =1;timeout s ESC \`.
I'm still considering how big the timeout limit should be, thinking about screen size / (10 + display speedup) in ms.

**Comment by avih** [comment]
> I'm still considering how big the timeout limit should be

In iTerm2 and my st patch it defaults to 200ms.

I think the practical considerations are that it's enough for big updates to complete (even over slow-ish connection) while feeling reasonably responsive if you abort the program, e.g. with ctrl-c.

**Comment by mintty** [comment]
> In iTerm2 and my st patch it defaults to 200ms.

Do you think it should be independent of screen size?
And should it be tuned by a general scrolling speedup factor (mintty option DisplaySpeedup)?
Also should it be disabled in "slow scroll" mode (DECSET 4 if implemented) as Brian suggested?

**Comment by avih** [comment]
The main thing to keep in mind is that this sequence is only a hint for the terminal. It can ignore it, or respect it up to any condition it considers a bail-out. The spec suggests some bail-outs, like resize, some modal things, obviously timeout, etc.

When used in practice, and to be effective, we're probably dealing with suspension of few ms at most, e.g. en editor or a pager refreshing the screen "atomically", some game or animation refreshing the screen, tmux when it needs to update a non-current pane ([hide the cursor,] move it elsewhere, output stuff, move back, [restore the cursor]) - without visible cursor flicker, etc.

The timeout will only be used with buggy or otherwise badly-behaving applications which don't end the suspension. Otherwise any reasonably sane application will end it as soon as the atomic update is complete - highly likely not more than few ms.

So I think the terminal should bail out as freely as it wants. I think terminal-resize and timeout are likely to be enough.

**Comment by mintty** [comment]
Refreshing the screen takes time depending on screen size, that's why I was asking. Not often relevant nowadays, perhaps, unless you were sitting at a serial connection or lagging network connection. But anyway, unless @BrianInglis objects, I'll ignore the screen size then.
"terminal ... can respect it up to any condition it considers": yes, but I'd prefer suggestions by people interested in such feature, not to decide on my own.
My modified proposal: use 200ms by default but allow the application to set up a different delay, limited to 800ms; 
this should also be enough to just ignore resizing actions.
Another question though: Should the timeout also reset the suspension delay or should it remain effective?

**Comment by avih** [comment]
> any reasonably sane application will end it as soon as the atomic update is complete - highly likely not more than few ms.

By the way, for this reason, I don't think the application should be allowed control the timeout. How does it know in advance if it needs 10ms or 50ms to push the next atomic output?

The RFC allows the application to extend the timeout it as much as it wants - by sending more BSU sequences, which it can add occasionally to its output, for instance every 1K of output. But even that's not required IMO.

It just send BSU, does output, send ESU, and that's it. The terminal, which serves the user, decides when enough is enough. Not the application.

**Comment by scandum** [comment]
> I'm still considering how big the timeout limit should be, thinking about screen size / (10 + display speedup) in ms.

Would it be possible to update in 1 second intervals if someone supplies something like 10000? Just keep subtracting 1000 until it reaches 0. 

It's also an option to allow supplying the update interval as the second argument.

Also, does mintty instantly update on every &q code it receives, or only if the timeout is set to 0?

**Comment by BrianInglis** [comment]
I know from human/system performance/responsiveness studies that ~100ms is the limit of human tolerance - more and the terminal/system will likely be seen as unresponsive, slow, hung, or dead, and may attract a `<ctrl-C>` *INTR* or `<ctrl-\>` *QUIT* from impatient users. This corresponds to 96 chars @ 9600b/s so was long enough for line end characters to be seen on normal speed and width I/O, pass the data, and reset the timeout.

**Comment by BrianInglis** [comment]
> slow scroll is not supported (I once tried to implement it but failed miserably; just getting another idea how it could be done, but does anybody really want it?).

It's just another feature for completeness so it is up to the implementer. 
This feature was introduced because the vt100 could display characters as fast as local serial ports could send them at 19200b/s == 24*80B/s
People appear to have been used to reading system output as it was displayed. 
Earlier video terminals may just have been higher speed electronic emulations of mechanical ttys, which were clocked by the serial data, so limited to at most 10c/s, or less when performing carriage or platen movement. 

Slow scroll rendered and scrolled each scan line separately, and I thought that appeared to operate like ~2400b/s, equivalent to about 3line/s, but the quote below suggests a maximum rate up to double that (possibly only 5line/s for Europe/Asia standard 50Hz AC frequency, as screen refresh rate often varies with power line frequency?) 
The manual says the terminal could operate at 4800b/s [full speed] without XON/XOF controls or delay padding, but features such as slow scroll require XON/XOF support. 

From vt100.net docs - [Scrolling Mode](https://vt100.net/docs/vt100-ug/chapter3.html#DECSCLM):
> The set state causes scrolls to be "smooth" at a maximum rate of six lines per second.

**Comment by mintty** [comment]
I plan to go with an implementation according to this description (for the wiki):

## Synchronous update ##

A pair of Begin/End Synchronous Update DCS sequences suspends the output 
between them in order to be updated to the screen synchronously.
The purpose is that applications can control atomic screen update, 
in order to avoid screen flickering in certain situations of display update.

| **sequence**      | **function**                                        |
|:------------------|:----------------------------------------------------|
| `^[P=1s^[\`       | suspend screen update for 150ms                     |
| `^[P=1;`_N_`s^[\` | suspend screen update for _N_ ms, max 420ms          |
| `^[P=2s^[\`       | update screen (flush output), end update suspending |

**Comment by scandum** [comment]
Will mintty accept `\a` as an alternative for `\e\` ?

**Comment by mintty** [comment]
No, that's only an alternative for OSC sequences.

**Comment by mintty** [comment]
Released 3.5.0.

**Comment by j4james** [comment]
@mintty FYI, some terminals are using private mode 2026 rather than the DCS sequence for synchronized updates. The benefit being that DECSET is unlikely to break terminals that don't support the sequence (which can be a problem with DCS). So you may also want to consider supporting DECSET 2026 as an alias for the DCS sequences. DECSET begins the synchronized update, and DECRST ends it.

**Comment by mintty** [comment]
Thanks for the hint. Are there any references? Which terminals?

**Comment by j4james** [comment]
The subject was discussed in the terminal-wg workgroup, and it was agreed that a mode was preferable to DCS, but no spec was ever drawn up. The closest to a spec is probably the thread starting here:

https://gitlab.freedesktop.org/terminal-wg/specifications/-/merge_requests/2#note_579597

However, note that the initial mode number proposed in that thread was 2001, but it was changed to 2026 because of a conflict with XTerm.

I know Contour definitely supports it (it's mentioned in the readme [here](https://github.com/christianparpart/contour/blob/master/README.md)), and both the iTerm2 and Windows Terminal devs have agreed to support it (see the discussion [here](https://github.com/microsoft/terminal/issues/8331#issuecomment-730471059)). It's also clear from that discussion that Kitty definitely _won't_ support it, but that's not surprising.

**Comment by mintty** [comment]
I've tentatively implemented this for mintty, quite easy.
However, I'm hesitating to commit the change as the DECSET mode isn't yet clearly documented anywhere, just in a merge request of the terminal working group, and although one could guess not clearly stating which way the mode is interpreted to sync (on DECSET on or off?)...

**Comment by j4james** [comment]
Based on that thread I linked to in terminal-wg, and the Contour implementation, the mappings should be:

`CSI ?2026h` -> BSU (begin synchronized update)
`CSI ?2026l` -> ESU (end synchronized update)

**Comment by j4james** [comment]
@mintty Thanks for adding the DECSET 2026 support. I haven't tested it yet, but I was looking at the commit and I noticed you don't force an update when the mode is disabled the way you do with the DCS sequence. Is it not necessary in this case? Sorry if this is a dumb question, but I just want to make sure it wasn't an oversight.

**Comment by mintty** [comment]
It *was* an oversight, thanks for noting.

**Comment by christianparpart** [comment]
I came here to see if you are supporting it and how. @mintty i am glad you also decided for SM/RM?2026. Many thanks. ;)

**Comment by mintty** [comment]
Released 3.5.1.
